src/compute/src/compute_state/peek_stash.rs

```diff
@@ -0,0 +1,245 @@
+    /// later.
+    stashed_rows: Option<Result<Vec<(Row, NonZeroI64)>, String>>,
+    /// The result of the background task, eventually.
+    pub result: oneshot::Receiver<(PeekResponse, Duration)>,
+    /// A background task that's responsible for producing the peek results.
+    /// If we're no longer interested in the results, we abort the task.
+    _abort_handle: AbortOnDropHandle<()>,
+}
+
+impl StashPeekResponse {
+    pub fn start_upload(
+        persist_clients: Arc<PersistClientCache>,
+        persist_location: &PersistLocation,
+        peek: Peek,
+        mut trace_bundle: TraceBundle,
+    ) -> Self {
+        let (rows_tx, rows_rx) = tokio::sync::mpsc::channel(10);
+        let (result_tx, result_rx) = oneshot::channel::<(PeekResponse, Duration)>();
+
+        let persist_clients = Arc::clone(&persist_clients);
+        let persist_location = persist_location.clone();
+
+        let peek_uuid = peek.uuid;
+        let relation_desc = peek.result_desc.clone();
+
+        let oks_handle = trace_bundle.oks_mut();
+
+        let peek_iterator = peek_result_iterator::PeekResultIterator::new(
+            peek.target.id().clone(),
+            peek.map_filter_project.clone(),
+            peek.timestamp,
+            peek.literal_constraints.clone(),
+            oks_handle,
+        );
+
+        let task_handle = mz_ore::task::spawn(|| "compute::stash_peek_response", async move {
+            let start = Instant::now();
+
+            let result = Self::do_upload(
+                &persist_clients,
+                persist_location,
+                peek.uuid,
+                relation_desc,
+                rows_rx,
+            )
+            .await;
+
+            let result = match result {
+                Ok(peek_response) => peek_response,
+                Err(e) => PeekResponse::Error(e.to_string()),
+            };
+            match result_tx.send((result, start.elapsed())) {
+                Ok(()) => {}
+                Err((_result, elapsed)) => {
+                    debug!(duration = ?elapsed, "dropping result for cancelled peek {}", peek_uuid)
+                }
+            }
+        });
+
+        Self {
+            peek,
+            peek_iterator: Some(peek_iterator),
+            rows_tx: Some(rows_tx),
+            stashed_rows: None,
+            result: result_rx,
+            _abort_handle: task_handle.abort_on_drop(),
+        }
+    }
+
+    async fn do_upload(
+        persist_clients: &PersistClientCache,
+        persist_location: PersistLocation,
+        peek_uuid: Uuid,
+        relation_desc: RelationDesc,
+        mut rows_rx: tokio::sync::mpsc::Receiver<Result<Vec<(Row, NonZeroI64)>, String>>,
+    ) -> Result<PeekResponse, String> {
+        let client = persist_clients
+            .open(persist_location)
+            .await
+            .map_err(|e| e.to_string())?;
+
+        let shard_id = format!("s{}", peek_uuid);
+        let shard_id = ShardId::try_from(shard_id).expect("can parse");
```

Comment:
Huh, are we allowed to do that? :D I guess nothing goes wrong, but `ShardId` doesn't explicitly document its string format, afaict, so it feels a bit like we are relying on implementation details.

Reply:
Yes, on your first part, and also the second paragraph!